lib/sqlalchemy/sql/elements.py

```diff
@@ -3423,24 +3423,26 @@ def __init__(
             for (c, r) in new_whens
         ]
 
-        if whenlist:
-            type_ = whenlist[-1][-1].type
-        else:
-            type_ = None
-
         if value is None:
             self.value = None
         else:
             self.value = coercions.expect(roles.ExpressionElementRole, value)
 
-        self.type = cast(_T, type_)
-        self.whens = whenlist
-
         if else_ is not None:
             self.else_ = coercions.expect(roles.ExpressionElementRole, else_)
         else:
             self.else_ = None
 
+        type_ = next(
+            (
```

Comment:
**Federico Caselli** (CaselIT) wrote:

A couple of comments.
- Since before it was using the last value, maybe it better to use `reversed(self.whens)`? If this is 2.1 only then it's probably fine like this too (and should be easier do document too), but I think reversed could be better if we backport to 2.0
- maybe instead of doing `isinstance(then, Null)` it could use `if not then.type._isnull` to obtain the fist non-null type?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5084